### PR TITLE
fetch: warn when using patch method

### LINKED ISSUE #2294: Undici `fetch` should at least warn on `method: "patch"` since it does not normalize like `http`
### What is the problem this feature will solve?

I spent the better part of the morning debugging why outgoing requests starting failing for only our PATCH requests after moving to `fetch` from `http` for one of our internal tools.

It ultimately was because undici ([to be fair, following the spec](https://fetch.spec.whatwg.org/#concept-method-normalize)) does not normalize `patch` to `PATCH` [unlike other common HTTP verbs](https://github.com/nodejs/node/blob/c19b2a761cca4cb6b811e4549e573b28dba1e99b/deps/undici/src/lib/fetch/util.js#L697-L701).

[But this is not the behavior of http, which normalizes `patch` to `PATCH` (all verbs, actually)
](https://github.com/nodejs/node/blob/c19b2a761cca4cb6b811e4549e573b28dba1e99b/lib/_http_client.js#L214)

Node itself will reject a request where the verb is `patch`, as all other servers and services I have tested.

### What is the feature you are proposing to solve the problem?

I propose node undici fetch emit a warning if the method "patch" is received as an option, informing the user that this HTTP method will not be normalized. 

This would operate similar to how node emits (used to emit?) a warning if a promise rejection is not handled.

### What alternatives have you considered?

I would frankly much rather prefer the old `http` behavior that normalizes all verbs, or at the very least makes an exception for PATCH, as the decision of a service to support `patch` with different semantics than `PATCH` strikes me as an extremely uncommon use case.

That said, I understand undici in particular favors specification compliance, so emitting a warning will be helpful in the short term as more folks like me move projects over to the native API.

### PR DESCRIPTION
**Pull request by KhafraDev**
Fixes #2294

### CODE REVIEW

**Review by metcoder95 — COMMENTED** [review]
See and understand the intention here, but sadly still not convinced with doing this. I'd still prefeer improve the documentation with a proper FAQ around this.

But not intention to block if we agree its the best way to go

**Comment by KhafraDev** [comment]
The issue is that it looks like an issue with undici, especially when moving from node-fetch where it works. The spec behavior is inconsistent here, and since node-fetch -> undici is probably a fairly common transition, a warning seems like the best to me. Plus, when using `patch`, the request might "succeed" but have subtle behavioral bugs later on (such as when parsing the body) - it being an issue with the method isn't apparent at all.

cc @ronag

**Review by mcollina — CHANGES_REQUESTED** [review]
I don't understand this. Why don't we capitalize all methods?

**Comment by KhafraDev** [comment]
@mcollina 

> To normalize a [method](https://fetch.spec.whatwg.org/#concept-method), if it is a [byte-case-insensitive](https://infra.spec.whatwg.org/#byte-case-insensitive) match for `DELETE`, `GET`, `HEAD`, `OPTIONS`, `POST`, or `PUT`, [byte-uppercase](https://infra.spec.whatwg.org/#byte-uppercase) it.

https://fetch.spec.whatwg.org/#concept-method-normalize

File: test/fetch/issue-2294-patch-method.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+'use strict'
+
+const { test } = require('tap')
```
Comment (sosukesuzuki) [inline test/fetch/issue-2294-patch-method.js:3]: Why do we use `tap` rather than `node:test` here? (context: https://github.com/nodejs/undici/issues/2267)

Comment (KhafraDev) [reply test/fetch/issue-2294-patch-method.js:3]: these tests use tap still

**Comment by mcollina** [comment]
I would recommend we deviate from the spec and just uppercase all of them.

**Comment by KhafraDev** [comment]
I don't agree, from discussions on the spec repo it seems like normalization was done for compatibility but is considered a mistake. It also could break legitimate use cases, in a rare case where someone might want `patch`.

> Now, in this case, there is spec divergence in Fetch from the HTTP spec, by having it normalize. This is non-ideal: it is not spec compliant (with HTTP). But, for better or worse, it reflects what was needed to avoid breaking legacy content because bugs in implementation quickly become assumptions by web developers. You can see that yourself, as you mention that despite HTTP specifying things are case-sensitive, developers have come to rely on assuming it was case insensitive due to historic bugs in browsers. That’s unfortunate, because it means you can write code for web browsers that won’t work in non-browser contexts, because only Web browsers have this “quirk”.

> [annevk] has rightly captured that, based on the currently available information, the current behavior is the right tradeoff, even if it may be surprising to some developers. The risks, of both breaking old code and of spec interoperability, are significant enough that even if absolute breakage isn’t quantified, as one of the foremost experts in making these tradeoff decisions, this is one where we know there is meaningful risk. The value proposition of making it easier for developers is important, but for such an issue like this, with available workarounds, it’s not appropriate to break old content and remove developer control (as normalizing would).

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
can you resolve the conflicts?

**Review by ronag — APPROVED** [review]

**Pull request merged**